### PR TITLE
feat: improve accuracy of p_error simulation in virtual circuits

### PR DESCRIPTION
**Pull request by umut-sahin**

### CODE REVIEW

File: concrete/numpy/representation/graph.py [hunk 1]
```diff
@@ -106,16 +108,31 @@ def evaluate(
                     if pred_node.operation != Operation.Input:
                         dtype = node.inputs[index].dtype
                         if isinstance(dtype, Integer):
-                            # this is not the real behavior of FHE
-                            # it's a simplified model, and it will be replaced at one point
+                            z = math.sqrt(2) * float(torch.erfinv(torch.tensor(1 - p_error)))
+                            p_error_per_error_value = {}
+
+                            error_value = 1
```
Comment (rudy-6-4) [inline concrete/numpy/representation/graph.py:114]: error_value is not clear maybe delta_error

File: concrete/numpy/representation/graph.py [hunk 2]
```diff
@@ -106,16 +108,31 @@ def evaluate(
                     if pred_node.operation != Operation.Input:
                         dtype = node.inputs[index].dtype
                         if isinstance(dtype, Integer):
-                            # this is not the real behavior of FHE
-                            # it's a simplified model, and it will be replaced at one point
+                            z = math.sqrt(2) * float(torch.erfinv(torch.tensor(1 - p_error)))
```
Comment (rudy-6-4) [inline concrete/numpy/representation/graph.py:111]: z is named fatal_noise in the optimizer
it could be named noise_delta_error_1

File: concrete/numpy/representation/graph.py (hunk 1 above)
Comment (umut-sahin) [reply concrete/numpy/representation/graph.py:114]: How about `error_amount`?

File: concrete/numpy/representation/graph.py (hunk 2 above)
Comment (umut-sahin) [reply concrete/numpy/representation/graph.py:111]: Why not `fatal_noise` as well?

File: concrete/numpy/representation/graph.py (hunk 1 above)
Comment (umut-sahin) [reply concrete/numpy/representation/graph.py:114]: Or `error_size`?

Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: Went with `error_size`, felt the best. Let me know if you want something else.

File: concrete/numpy/representation/graph.py (hunk 2 above)
Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: Renamed to `fatal_noise`. Let me know if you want something else.

Comment (andrei-stoian-zama) [reply concrete/numpy/representation/graph.py]: you're using torch here but in other issues there was discussion about removing the torch dependency, does this code lock you in to torch ?

Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: Actually, it lacks `erfcinv` which is needed for the comment below. So we'll probably rewrite this to use `scipy` directly.

**Review by andrei-stoian-zama — COMMENTED** [review]
Complex equations merit a clear textual explanation :smile: 

Can you please explain the computation that you implemented ?

File: concrete/numpy/representation/graph.py (hunk 2 above)
Comment (aquint-zama) [reply concrete/numpy/representation/graph.py]: yes plz, let's try to remove torch if possible

Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: Done.

[the remark this answers is not in the document]
Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: Done.

**Comment by umut-sahin** [comment]
> Complex equations merit a clear textual explanation smile
> 
> Can you please explain the computation that you implemented ?

I've added the PDF created by @samcrx to the docs and referred to it.

File: concrete/numpy/representation/graph.py [hunk 3]
```diff
@@ -106,16 +110,34 @@ def evaluate(
                     if pred_node.operation != Operation.Input:
                         dtype = node.inputs[index].dtype
                         if isinstance(dtype, Integer):
-                            # this is not the real behavior of FHE
-                            # it's a simplified model, and it will be replaced at one point
+                            # see https://github.com/zama-ai/concrete-numpy/blob/main/docs/_static/p_error_simulation.pdf  # noqa: E501  # pylint: disable=line-too-long
```
Comment (andrei-stoian-zama) [inline concrete/numpy/representation/graph.py:113]: this computation may be long, maybe an idea would to use a kind of cache ? When the circuit is compiled, compute the delta error probabilities for each TLU (if `p_error` is not the same for all TLU because the user used `global_p_error`) or for all TLUs (when the `p_error` is used). 

then here just retrieve the computation result using the `p_error` which seems to be the only parameter that influences this computation

**Review by andrei-stoian-zama — COMMENTED** [review]
A suggestion to cache this so not to evaluate the error computation every time


I can't tell if the computation is correct following the PDF but it is explained quite well in the PDF. Does the implementation seem ok @samcrx ?

I ran the code to simulate the pbs error, I get this:


delta_p_error

does this seem ok?

File: concrete/numpy/representation/graph.py (hunk 3 above)
Comment (andrei-stoian-zama) [reply concrete/numpy/representation/graph.py:115]: nice!!

**Review by rudy-6-4 — DISMISSED** [review]

File: docs/linux.dependency.licenses.txt [hunk 4]
```diff
@@ -15,6 +15,7 @@
  packaging                 23.0        Apache Software License; BSD License
  pyparsing                 3.0.9       MIT License
  python-dateutil           2.8.2       Apache Software License; BSD License
+ scipy                     1.7.3       BSD License
  six                       1.16.0      MIT License
  torch                     1.13.1      BSD License
```
Comment (rudy-6-4) [inline docs/linux.dependency.licenses.txt:20]: should be removed ?

File: pyproject.toml [hunk 5]
```diff
@@ -44,6 +44,7 @@ numpy = "^1.21.0"
 Pillow = "^9.0.0"
 concrete-compiler = "^0.23.4"
 torch = "^1.13.1"
```
Comment (rudy-6-4) [inline pyproject.toml:46]: should be removed ?

File: docs/linux.dependency.licenses.txt (hunk 4 above)
Comment (umut-sahin) [reply docs/linux.dependency.licenses.txt:20]: No, it's used for convolutions and maxpool.

File: pyproject.toml (hunk 5 above)
Comment (umut-sahin) [reply pyproject.toml:46]: No, it's used for convolutions and maxpool.

**Comment by samcrx** [comment]
@andrei-stoian-zama  On your graph it's weird that we don't have the same frequency for +1 and -1 (p_error=0.5). There must be a mistake somewhere 🤔

File: concrete/numpy/representation/graph.py [hunk 6]
```diff
@@ -106,16 +112,46 @@ def evaluate(
...
+                                p_error_per_error_size = {}
+
+                                error_size = 1
+                                last_p_error = p_error
+                                while last_p_error > 2 ** (-80):
+                                    new_fatal_noise = (2**error_size) * fatal_noise
+                                    new_p_error = scipy.special.erfc(new_fatal_noise / math.sqrt(2))
+
+                                    p_error_per_error_size[error_size] = last_p_error - new_p_error
+
```
Comment (samcrx) [inline concrete/numpy/representation/graph.py:129]: I think it should be

```python
new_fatal_noise = (2*error_size + 1) * fatal_noise
new_p_error = scipy.special.erfc(new_fatal_noise / math.sqrt(2))
p_error_per_error_size[error_size] =  new_p_error - last_p_error
```

to match with the formula in the PDF

also, `fatal_noise` and `new_fatal_noise` are not well named, it should be `std_score` and `new_std_score` (std=standard)

**Review by samcrx — CHANGES_REQUESTED** [review]

File: concrete/numpy/representation/graph.py (hunk 6 above)
Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: I did the renaming and `+ 1` after `2**error_size`, but changing `last_p_error - new_p_error` to `new_p_error - last_p_error` makes the probabilities negative so I kept it as is.

Comment (umut-sahin) [reply concrete/numpy/representation/graph.py]: Done with some other changes.

**Review by rudy-6-4 — APPROVED** [review]

**Review by samcrx — APPROVED** [review]
